libraries/ESP8266WebServer/examples/FSBrowser/FSBrowser.ino

```diff
@@ -226,29 +302,324 @@ void handleFileList() {
+   return the path of the closest parent still existing
+*/
+String lastExistingParent(String path) {
+  while (path != "" && !fileSystem->exists(path)) {
+    if (path.lastIndexOf("/") > 0) {
+      path = path.substring(0, path.lastIndexOf("/"));
+    } else {
+      path = String();  // No slash => the top folder does not exist
+    }
+  }
+  DBG_OUTPUT_PORT.println(String("Last existing parent: ") + path);
+  return path;
+}
+
+/*
+   Handle the creation/rename of a new file
+   Operation      | req.responseText
+   ---------------+--------------------------------------------------------------
+   Create file    | parent of created file
+   Create folder  | parent of created folder
+   Rename file    | parent of source file
+   Move file      | parent of source file, or remaining ancestor
+   Rename folder  | parent of source folder
+   Move folder    | parent of source folder, or remaining ancestor
+*/
+void handleFileCreate() {
+  if (!fsOK) {
+    return returnFail("FS INIT ERROR");
+  }
+
+  String path = server.arg("path");
+  if (path == "") {
+    return returnFail("MISSING PATH ARG");
+  }
+
+#ifdef USE_SPIFFS
+  if (getFileError(path).length() > 0) {
+    return returnFail("INVALID FILENAME");
+  }
+#endif
+
+  if (path == "/") {
+    return returnFail("BAD PATH");
+  }
+  if (fileSystem->exists(path)) {
+    return returnFail("PATH FILE EXISTS");
+  }
+
+  String src = server.arg("src");
+  if (src == "") {
+    // No source specified: creation
+    DBG_OUTPUT_PORT.println(String("handleFileCreate: ") + path);
+    if (path.endsWith("/")) {
+      // Create a folder
+      path.remove(path.length() - 1);
+      if (!fileSystem->mkdir(path)) {
+        return returnFail("MKDIR FAILED");
+      }
+    } else {
+      // Create a file
+      File file = fileSystem->open(path, "w");
+      if (file) {
+        file.write((const char *)0);
+        file.close();
+      } else {
+        return returnFail("CREATE FAILED");
+      }
+    }
+    if (path.lastIndexOf("/") > -1) {
+      path = path.substring(0, path.lastIndexOf("/"));
+    }
+    returnOKWithMsg(path);
+  } else {
+    // Source specified: rename
+    if (src == "/") {
+      return returnFail("BAD SRC");
+    }
+    if (!fileSystem->exists(src)) {
+      return returnFail("SRC FILE NOT FOUND");
+    }
+
+    DBG_OUTPUT_PORT.println(String("handleFileCreate: ") + path + " from " + src);
+
+    if (path.endsWith("/")) {
+      path.remove(path.length() - 1);
+    }
+    if (src.endsWith("/")) {
+      src.remove(src.length() - 1);
+    }
+    if (!fileSystem->rename(src, path)) {
+      return returnFail("RENAME FAILED");
+    }
+    returnOKWithMsg(lastExistingParent(src));
+  }
+}
+
+
+/*
+   Delete the file or folder designed by the given path.
+   If it's a file, delete it.
+   If it's a folder, delete all nested contents first then the folder itself
+*/
+void deleteRecursive(String path) {
```

Comment:
Been thinking about this for a few hours, but I'm not sure at all how to do it :-(.
I guess one way or another, I'd have to push the "yet-to-process" items on a custom-made stack, so I wonder if it's worth it... 
Now regarding the passing by reference, the only solution I can see would be to replace the "path" String by a linked list of pointers to each level.
But again, I fear that such an improvement, altough required for a high quality project, would make code more complex and less legible for an example meant to demonstrate usage of the web server library...

Reply:
Alright, leave it as is for now, but because this is an example please add a comment above saying something along the lines of "This function uses recursion, don't do this at home or you'll blow up the stack"